### ISSUE TITLE
Regression in issue #8496 with canvas backend

### ISSUE THREAD

**Issue by janpe2** [mozilla/pdf.js#9398]
Issue #8496 is about the SVG backend but now also the **canvas** backend has problems with the same PDF. I have tested a few versions: 1.7.225 and 1.8.188 have no problems but newer versions like 1.9.583 (built-in viewer in Firefox), 1.10.96 and 2.0.288 (current demo viewer) don't render the file correctly.

Attach (recommended) or Link to PDF file here: 
PDF attached in issue #8496

Configuration:
- Web browser and its version: Firefox 57.0.4
- Operating system and its version: Windows 7
- PDF.js version: see above
- Is a browser extension:

Steps to reproduce the problem:
1. Open the mentioned PDF with the default canvas backend enabled.

What is the expected behavior? 
Version 1.7.225 renders all texts fine:
![kuva](https://user-images.githubusercontent.com/26090208/35292820-66505e36-007a-11e8-8a14-4d023c3e9762.png)

What went wrong? 
Current demo viewer 2.0.288 has some parts of texts missing:
![kuva](https://user-images.githubusercontent.com/26090208/35292863-82b93ee4-007a-11e8-9c5b-e98c29d4901b.png)

**Comment by brendandahl** [comment]
Could you do a `git bisect` to narrow down the regression?

**Comment by Snuffleupagus** [comment]
> Could you do a git bisect to narrow down the regression?

It seems that PR #8792 is the culprit here.

**Comment by janpe2** [comment]
[issue9398_reduced.pdf](https://github.com/mozilla/pdf.js/files/1657815/issue9398_reduced.pdf)

I extracted two inline images from the original PDF. The page should show a gray and a black image but PDF.js renders two gray images. The data of the two images are equal but their color spaces are different.

**Comment by brendandahl** [comment]
> Edit2: Given that parsing/checking the Dicts aren't free, how that should then be implemented without regressing issue #2618

I looked quickly at the #2618 PDF and the one in this issue and both dictionaries are inline so we're already parsing them: https://github.com/mozilla/pdf.js/blob/master/src/core/parser.js#L374-L385

**Comment by Snuffleupagus** [comment]
> I looked quickly at the #2618 PDF and the one in this issue and both dictionaries are inline so we're already parsing them: https://github.com/mozilla/pdf.js/blob/master/src/core/parser.js#L374-L385

Sorry, I didn't express myself clearly above!

What I meant is that we need to somehow serialize/stringify/hash the contents of those Dictionaries, and use that when building the inline image `cacheKey`, in order to fix this bug (and prevent future similar ones). However, in #2618 there's around 160000 inline images (if memory serves), and when I tried implementing (a naive) hashing of the Dictionaries I ended up regressing that particular PDF file by `> 100 %`.
